### PR TITLE
check job_status before BatchOperator execute in deferrable mode

### PR DESCRIPTION
**Pull request by Lee-W**
While running a batch job in deferrable mode, the condition might already be met before we defer the task into the trigger. This PR intends to check the batch status before deferring the task to trigger.

---
**^ Add meaningful description above**
Read the **[Pull Request Guidelines](https://github.com/apache/airflow/blob/main/CONTRIBUTING.rst#pull-request-guidelines)** for more information.
In case of fundamental code changes, an Airflow Improvement Proposal ([AIP](https://cwiki.apache.org/confluence/display/AIRFLOW/Airflow+Improvement+Proposals)) is needed.
In case of a new dependency, check compliance with the [ASF 3rd Party License Policy](https://www.apache.org/legal/resolved.html#category-x).
In case of backwards incompatible changes please leave a note in a newsfragment file, named `{pr_number}.significant.rst` or `{issue_number}.significant.rst`, in [newsfragments](https://github.com/apache/airflow/tree/main/newsfragments).

### CODE REVIEW

**Review by dirrao — APPROVED** [review]
LGTM.

**Comment by vincbeck** [comment]
> > If the condition is already met and the task is deferred, what happens? My understanding is the task gets executed successfully. So the gain here is just time? Instead of deferring the task for nothing (because the condition is already met), we return it directly. Unless there is a very good reason to do so I am not really in favor of this change because it basically duplicates what the trigger is already doing but in the operator itself.
> 
> Yes, I think the main gain here is time as we avoid unnecessary serialization and deserialization. Another thing is for consistency. We already have this behavior for some of the operators
> 
> https://github.com/apache/airflow/blob/223a98420bbdfd0256e58f4e9aa3a06bda914c8a/airflow/providers/amazon/aws/sensors/emr.py#L647-L658
> 
> https://github.com/apache/airflow/blob/223a98420bbdfd0256e58f4e9aa3a06bda914c8a/airflow/providers/amazon/aws/operators/emr.py#L578-L597
> 
> > If we go down that path, I guess this logic can apply to any deferrable operator, and as such, such logic should be copied over to all deferrable operators
> 
> Yes, I think that's something we should do eventually

I see, I dont necessarily agree but I'll wait others to comment, I might be the only grumpy one here :)

**Comment by potiuk** [comment]
> I see, I dont necessarily agree but I'll wait others to comment, I might be the only grumpy one here :)

Just a little grumpy, I think :)

> Yes, I think that's something we should do eventually

Yes I agree we might want to do it. Eventually, It's an optimisation and we should continue doing it. We've had similar cases in the past - for example when we optimized EmptyOperator. It's hard to have an enforceable rule here though,  so I'd say ad-hoc optimisation like this one is the best approach.

**Review by potiuk — APPROVED** [review]

File: airflow/providers/amazon/aws/operators/batch.py [hunk 1]
```diff
@@ -230,36 +230,55 @@ def hook(self) -> BatchClientHook:
...
+                raise AirflowException("AWS Batch job - job_id was not found")
+
+            job = self.hook.get_job_description(self.job_id)
+            job_status = job.get("status")
+            if job_status == self.hook.SUCCESS_STATE:
+                msg = f"{self.job_id} was completed successfully"
+                self.log.info(msg)
+                return self.job_id
+            elif job_status == self.hook.FAILURE_STATE:
+                raise AirflowException(f"{self.job_id} failed")
```
Comment (o-nikolas) [inline airflow/providers/amazon/aws/operators/batch.py]: Should we add this state inspection/logging/handling into `execute_complete` as well?

**Review by o-nikolas — COMMENTED** [review]
I also think this is a bit redundant. I like the time savings, but it complicates the code a bit to have result checking/hanlding in multiple places. I think if we move forward with this we should do our best to abstract that stuff into functions that are used in both the defer and non-defer paths.

File: airflow/providers/amazon/aws/operators/batch.py (hunk 1 above)
Comment (Lee-W) [reply airflow/providers/amazon/aws/operators/batch.py]: I think the state inspection should happen in triggerer. The logging's already in `execute_complete`, but yes, I should refine it a bit to make it consistent. Thanks for reminding me!

Comment (Lee-W) [reply airflow/providers/amazon/aws/operators/batch.py]: @o-nikolas, I tried to refine it a bit. Please let me know if it looks good to you. Thanks!

**Review by phanikumv — APPROVED** [review]

**Comment by phanikumv** [comment]
> I also think this is a bit redundant. I like the time savings, but it complicates the code a bit to have result checking/hanlding in multiple places. I think if we move forward with this we should do our best to abstract that stuff into functions that are used in both the defer and non-defer paths.

Merging this for now. @Lee-W can you implement the abstraction part in a separate PR

**Pull request merged**